# Superstore Sales SQL Analysis
### Steps Covered:
1. Load Dataset
2. Explore Table (schema, sample data, row count)
3. Apply WHERE Filters
4. GROUP BY Aggregations
5. Sort & Limit Results
6. Use Cases (monthly trends, top customers, duplicates)
7. Data Quality Checks

---
## Setup - Import Libraries (SQLite)

In [24]:
import sqlite3
import pandas as pd

---
## Step 1 - Load Dataset into SQL Database
> This step reads the CSV file and creates a SQL table called `db`.

In [25]:
CSV_PATH = 'Sample_Superstore_Underscore.csv'

df = pd.read_csv(CSV_PATH)

conn = sqlite3.connect('superstore.db')

df.to_sql('db', conn, if_exists='replace', index=False)

print("Data loaded into SQLite table")

Data loaded into SQLite table


---
## Step 2 - Explore the Table

### 2a. Total Rows in Table
> Counts the total number of records imported.

In [26]:
pd.read_sql_query("""
    SELECT COUNT(*) AS total_rows
    FROM db
""", conn)

,total_rows
0,9994


### 2b. Sample Data (LIMIT 10)
> Displays the first 10 records for quick inspection.

In [27]:
pd.read_sql_query("""
    SELECT *
    FROM db
    LIMIT 10
""", conn)

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,postal_code,region,product_id,category,sub_category,product_name,sales,quantity,discount,profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
5,6,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.8600,7,0.00,14.1694
6,7,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.2800,4,0.00,1.9656
7,8,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.1520,6,0.20,90.7152
8,9,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by S...,18.5040,3,0.20,5.7825
9,10,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9000,5,0.00,34.4700


### 2c. Schema / Structure
> Shows all column names and their data types.

In [28]:
pd.read_sql_query("""
PRAGMA table_info('db');
""", conn)

,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,order_id,TEXT,0,None,0
2,2,order_date,TEXT,0,None,0
3,3,ship_date,TEXT,0,None,0
4,4,ship_mode,TEXT,0,None,0
5,5,customer_id,TEXT,0,None,0
6,6,customer_name,TEXT,0,None,0
7,7,segment,TEXT,0,None,0
8,8,country,TEXT,0,None,0
9,9,city,TEXT,0,None,0


---
## Step 3 - Apply WHERE Filters

### 3a. Filter by Region = 'West'
> Retrieves all sales records from the West region.

In [29]:
result = pd.read_sql_query("""
    SELECT *
    FROM db
    WHERE region = 'West'
""", conn)

print(f"Total rows (West region): {len(result)}")
result.head(10)

Total rows (West region): 3203


,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,postal_code,region,product_id,category,sub_category,product_name,sales,quantity,discount,profit
0,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.620,2,0.0,6.8714
1,6,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.860,7,0.0,14.1694
2,7,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.280,4,0.0,1.9656
3,8,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.152,6,0.2,90.7152
4,9,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by S...,18.504,3,0.2,5.7825
5,10,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.900,5,0.0,34.4700
6,11,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-TA-10001539,Furniture,Tables,Chromcraft Rectangular Conference Tables,1706.184,9,0.2,85.3092
7,12,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002033,Technology,Phones,Konftel 250 Conference phone - Charcoal black,911.424,4,0.2,68.3568
8,14,CA-2016-161389,12/5/2016,12/10/2016,Standard Class,IM-15070,Irene Maddox,Consumer,United States,Seattle,...,98103,West,OFF-BI-10003656,Office Supplies,Binders,Fellowes PB200 Plastic Comb Binding Machine,407.976,3,0.2,132.5922
9,18,CA-2014-167164,5/13/2014,5/15/2014,Second Class,AG-10270,Alejandro Grove,Consumer,United States,West Jordan,...,84084,West,OFF-ST-10000107,Office Supplies,Storage,Fellowes Super Stor/Drawer,55.500,2,0.0,9.9900


### 3b. Filter by Category = 'Furniture'
> Displays only records from the Furniture category.

In [30]:
result = pd.read_sql_query("""
    SELECT *
    FROM db
    WHERE category = 'Furniture'
""", conn)

print(f"Total rows (Furniture): {len(result)}")
result.head(10)

Total rows (Furniture): 2121


,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,postal_code,region,product_id,category,sub_category,product_name,sales,quantity,discount,profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
3,6,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.8600,7,0.00,14.1694
4,11,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-TA-10001539,Furniture,Tables,Chromcraft Rectangular Conference Tables,1706.1840,9,0.20,85.3092
5,24,US-2017-156909,7/16/2017,7/18/2017,Second Class,SF-20065,Sandra Flanagan,Consumer,United States,Philadelphia,...,19140,East,FUR-CH-10002774,Furniture,Chairs,"Global Deluxe Stacking Chair, Gray",71.3720,2,0.30,-1.0196
6,25,CA-2015-106320,9/25/2015,9/30/2015,Standard Class,EB-13870,Emily Burns,Consumer,United States,Orem,...,84057,West,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,1044.6300,3,0.00,240.2649
7,28,US-2015-150630,9/17/2015,9/21/2015,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,...,19140,East,FUR-BO-10004834,Furniture,Bookcases,"Riverside Palais Royal Lawyers Bookcase, Royal...",3083.4300,7,0.50,-1665.0522
8,30,US-2015-150630,9/17/2015,9/21/2015,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,...,19140,East,FUR-FU-10004848,Furniture,Furnishings,"Howard Miller 13-3/4"" Diameter Brushed Chrome ...",124.2000,3,0.20,15.5250
9,37,CA-2016-117590,12/8/2016,12/10/2016,First Class,GH-14485,Gene Hale,Corporate,United States,Richardson,...,75080,Central,FUR-FU-10003664,Furniture,Furnishings,"Electrix Architect's Clamp-On Swing Arm Lamp, ...",190.9200,5,0.60,-147.9630


### 3c. Filter by Sales > 500
> Identifies high-value transactions.

In [31]:
result = pd.read_sql_query("""
    SELECT *
    FROM db
    WHERE sales > 500
""", conn)

print(f"Total rows (Sales > 500): {len(result)}")
result.head(10)

Total rows (Sales > 500): 1162


,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,postal_code,region,product_id,category,sub_category,product_name,sales,quantity,discount,profit
0,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
1,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
2,8,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.1520,6,0.20,90.7152
3,11,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-TA-10001539,Furniture,Tables,Chromcraft Rectangular Conference Tables,1706.1840,9,0.20,85.3092
4,12,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002033,Technology,Phones,Konftel 250 Conference phone - Charcoal black,911.4240,4,0.20,68.3568
5,17,CA-2014-105893,11/11/2014,11/18/2014,Standard Class,PK-19075,Pete Kriz,Consumer,United States,Madison,...,53711,Central,OFF-ST-10004186,Office Supplies,Storage,"Stur-D-Stor Shelving, Vertical 5-Shelf: 72""H x...",665.8800,6,0.00,13.3176
6,25,CA-2015-106320,9/25/2015,9/30/2015,Standard Class,EB-13870,Emily Burns,Consumer,United States,Orem,...,84057,West,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,1044.6300,3,0.00,240.2649
7,28,US-2015-150630,9/17/2015,9/21/2015,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,...,19140,East,FUR-BO-10004834,Furniture,Bookcases,"Riverside Palais Royal Lawyers Bookcase, Royal...",3083.4300,7,0.50,-1665.0522
8,36,CA-2016-117590,12/8/2016,12/10/2016,First Class,GH-14485,Gene Hale,Corporate,United States,Richardson,...,75080,Central,TEC-PH-10004977,Technology,Phones,GE 30524EE4,1097.5440,7,0.20,123.4737
9,39,CA-2015-117415,12/27/2015,12/31/2015,Standard Class,SN-20710,Steve Nguyen,Home Office,United States,Houston,...,77041,Central,FUR-BO-10002545,Furniture,Bookcases,"Atlantic Metals Mobile 3-Shelf Bookcases, Cust...",532.3992,3,0.32,-46.9764


### 3d. Filter by Date - Year 2017
> Extracts all orders placed during 2017.

In [32]:
result = pd.read_sql_query("""
SELECT *
FROM db
WHERE substr(order_date, -4) = '2017'
""", conn)

print(f"Total rows (Year 2017): {len(result)}")
result.head(10)

Total rows (Year 2017): 3312


,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,postal_code,region,product_id,category,sub_category,product_name,sales,quantity,discount,profit
0,13,CA-2017-114412,4/15/2017,4/20/2017,Standard Class,AA-10480,Andrew Allen,Consumer,United States,Concord,...,28027,South,OFF-PA-10002365,Office Supplies,Paper,Xerox 1967,15.552,3,0.2,5.4432
1,24,US-2017-156909,7/16/2017,7/18/2017,Second Class,SF-20065,Sandra Flanagan,Consumer,United States,Philadelphia,...,19140,East,FUR-CH-10002774,Furniture,Chairs,"Global Deluxe Stacking Chair, Gray",71.372,2,0.3,-1.0196
2,35,CA-2017-107727,10/19/2017,10/23/2017,Second Class,MA-17560,Matt Abelman,Home Office,United States,Houston,...,77095,Central,OFF-PA-10000249,Office Supplies,Paper,Easy-staple paper,29.472,3,0.2,9.9468
3,42,CA-2017-120999,9/10/2017,9/15/2017,Standard Class,LC-16930,Linda Cazamias,Corporate,United States,Naperville,...,60540,Central,TEC-PH-10004093,Technology,Phones,Panasonic Kx-TS550,147.168,4,0.2,16.5564
4,44,CA-2017-139619,9/19/2017,9/23/2017,Standard Class,ES-14080,Erin Smith,Corporate,United States,Melbourne,...,32935,South,OFF-ST-10003282,Office Supplies,Storage,"Advantus 10-Drawer Portable Organizer, Chrome ...",95.616,2,0.2,9.5616
5,72,CA-2017-114440,9/14/2017,9/17/2017,Second Class,TB-21520,Tracy Blumstein,Consumer,United States,Jackson,...,49201,Central,OFF-PA-10004675,Office Supplies,Paper,Telephone Message Books with Fax/Mobile Sectio...,19.050,3,0.0,8.7630
6,76,US-2017-118038,12/9/2017,12/11/2017,First Class,KB-16600,Ken Brennan,Corporate,United States,Houston,...,77041,Central,OFF-BI-10004182,Office Supplies,Binders,Economy Binders,1.248,3,0.8,-1.9344
7,77,US-2017-118038,12/9/2017,12/11/2017,First Class,KB-16600,Ken Brennan,Corporate,United States,Houston,...,77041,Central,FUR-FU-10000260,Furniture,Furnishings,"6"" Cubicle Wall Clock, Black",9.708,3,0.6,-5.8248
8,78,US-2017-118038,12/9/2017,12/11/2017,First Class,KB-16600,Ken Brennan,Corporate,United States,Houston,...,77041,Central,OFF-ST-10000615,Office Supplies,Storage,"SimpliFile Personal File, Black Granite, 15w x...",27.240,3,0.2,2.7240
9,85,US-2017-119662,11/13/2017,11/16/2017,First Class,CS-12400,Christopher Schild,Home Office,United States,Chicago,...,60623,Central,OFF-ST-10003656,Office Supplies,Storage,Safco Industrial Wire Shelving,230.376,3,0.2,-48.9549


---
## Step 4 - GROUP BY Aggregations

### 4a. Total Sales by Region
> West leads with highest total sales.

In [33]:
pd.read_sql_query("""
    SELECT region,
           ROUND(SUM(sales), 2) AS total_sales
    FROM db
    GROUP BY region
    ORDER BY total_sales DESC
""", conn)

,region,total_sales
0,West,725457.82
1,East,678781.24
2,Central,501239.89
3,South,391721.91


### 4b. Total Quantity by Category
> Shows total units sold per category.

In [34]:
pd.read_sql_query("""
    SELECT category,
           SUM(quantity) AS total_quantity
    FROM db
    GROUP BY category
    ORDER BY total_quantity DESC
""", conn)

,category,total_quantity
0,Office Supplies,22906
1,Furniture,8028
2,Technology,6939


### 4c. Average Sales by Category
> Technology has the highest average sales per transaction.

In [35]:
pd.read_sql_query("""
    SELECT category,
           ROUND(AVG(sales), 2) AS avg_sales
    FROM db
    GROUP BY category
    ORDER BY avg_sales DESC
""", conn)

,category,avg_sales
0,Technology,452.71
1,Furniture,349.83
2,Office Supplies,119.32


---
## Step 5 - Sort & Limit Results

### 5a. Top 10 Products by Sales
> Canon imageCLASS 2200 is the 1 revenue-generating product (~$61,599).

In [36]:
pd.read_sql_query("""
    SELECT product_name,
           ROUND(SUM(sales), 2) AS total_sales
    FROM db
    GROUP BY product_name
    ORDER BY total_sales DESC
    LIMIT 10
""", conn)

,product_name,total_sales
0,Canon imageCLASS 2200 Advanced Copier,61599.82
1,Fellowes PB500 Electric Punch Plastic Comb Bin...,27453.38
2,Cisco TelePresence System EX90 Videoconferenci...,22638.48
3,HON 5400 Series Task Chairs for Big and Tall,21870.58
4,GBC DocuBind TL300 Electric Binding System,19823.48
5,GBC Ibimaster 500 Manual ProClick Binding System,19024.50
6,Hewlett Packard LaserJet 3310 Copier,18839.69
7,HP Designjet T520 Inkjet Large Format Printer ...,18374.90
8,GBC DocuBind P400 Electric Binding System,17965.07
9,High Speed Automatic Electric Letter Opener,17030.31


### 5b. Top Categories by Profit
> Technology leads in profit, followed by Office Supplies. Furniture is the least profitable.

In [37]:
pd.read_sql_query("""
    SELECT category,
           ROUND(SUM(profit), 2) AS total_profit
    FROM db
    GROUP BY category
    ORDER BY total_profit DESC
""", conn)

,category,total_profit
0,Technology,145454.95
1,Office Supplies,122490.80
2,Furniture,18451.27


---
## Step 6 - Use Cases

### 6a. Monthly Profit Analysis
> Tracks how profit changes month by month across all years.

In [38]:
pd.read_sql_query("""
    SELECT
        strftime('%Y-%m', order_date) AS month,
        ROUND(SUM(profit), 2)           AS total_profit
    FROM db
    GROUP BY month
    ORDER BY month
""", conn)

,month,total_profit
0,None,286397.02


### 6b. Monthly Sales Analysis
> Evaluates monthly revenue trends over the full dataset period (2014–2017).

In [39]:
pd.read_sql_query("""
    SELECT
        strftime('%Y-%m', order_date) AS month,
        ROUND(SUM(sales), 2)            AS total_sales
    FROM db
    GROUP BY month
    ORDER BY month
""", conn)

,month,total_sales
0,None,2297200.86


### 6c. Top 10 Customers by Sales


In [40]:
pd.read_sql_query("""
    SELECT customer_name,
           COUNT(*)                  AS total_orders,
           ROUND(SUM(sales),  2)     AS total_sales,
           ROUND(SUM(profit), 2)     AS total_profit
    FROM db
    GROUP BY customer_name
    ORDER BY total_sales DESC
    LIMIT 10
""", conn)

,customer_name,total_orders,total_sales,total_profit
0,Sean Miller,15,25043.05,-1980.74
1,Tamara Chand,12,19052.22,8981.32
2,Raymond Buch,18,15117.34,6976.10
3,Tom Ashbrook,10,14595.62,4703.79
4,Adrian Barton,20,14473.57,5444.81
5,Ken Lonsdale,29,14175.23,806.86
6,Sanjit Chand,22,14142.33,5757.41
7,Hunter Lopez,11,12873.30,5622.43
8,Sanjit Engle,19,12209.44,2650.68
9,Christopher Conant,11,12129.07,2177.05


### 6d. Duplicate Order Detection
> Checks for order IDs appearing more than once.

In [41]:
result = pd.read_sql_query("""
    SELECT order_id,
           COUNT(*) AS count
    FROM db
    GROUP BY order_id
    HAVING COUNT(*) > 1
    ORDER BY count DESC
""", conn)

print(f"Order IDs with more than 1 row: {len(result)}")
result.head(10)

Order IDs with more than 1 row: 2471


,order_id,count
0,CA-2017-100111,14
1,CA-2017-157987,12
2,US-2016-108504,11
3,CA-2016-165330,11
4,US-2015-126977,10
5,CA-2016-105732,10
6,CA-2015-131338,10
7,US-2016-114013,9
8,US-2015-163433,9
9,CA-2017-140949,9


---
## Step 7 — Data Quality Checks

### 7a. Sales Equal to Zero
> Checks if any transaction has zero sales.

In [42]:
result = pd.read_sql_query("""
    SELECT *
    FROM db
    WHERE sales = 0
""", conn)

if len(result) == 0:
    print("No records with zero sales.")
else:
    print(f"Found {len(result)} records with zero sales")
    result

No records with zero sales.


### 7b. NULL Value Check
> Checks for missing values in critical columns.

In [43]:
result = pd.read_sql_query("""
    SELECT *
    FROM db
    WHERE order_id IS NULL
       OR customer_name IS NULL
""", conn)

if len(result) == 0:
    print("No NULL values in order_id or customer_name.")
else:
    print(f"Found {len(result)} records with NULL values")
    result

No NULL values in order_id or customer_name.


### 7c. Final Row Count Validation
> Confirms total rows after all operations. 

In [44]:
pd.read_sql_query("""
    SELECT COUNT(*) AS total_rows
    FROM db
""", conn)

,total_rows
0,9994


---
## Summary of Key Insights

| # | Query | Key Finding |
|---|-------|-------------|
| 1 | Total rows | 9,994 records in the dataset |
| 2 | Region filter | West has the most records (3,203) |
| 3 | High-value orders | 1,162 transactions with sales > $500 |
| 4 | Sales by region | West leads in total revenue |
| 5 | Avg sales by category | Technology has highest avg (~$452) |
| 6 | Top product | Canon imageCLASS 2200 (~$61,599 total sales) |
| 7 | Most profitable category | Technology ($145,454 profit) |
| 8 | Top customer | Sean Miller - highest sales but **negative profit** |
| 9 | Data quality | No NULLs, no zero-sales records,dataset is clean |
| 10 | Duplicates | Expected - one order ID can have multiple product rows |
